File: 9.Caching, the secret behind it all/code/typescript/104_session_management_with_redis_python_1.ts
import Redis from 'ioredis';

export interface Product {
    id: string;
    name: string;
    price: number;
}

const redis = new Redis({
    host: 'localhost',
    port: 6379,
});

// GetProduct implements Cache-Aside (Lazy) caching.
// 1. Check Redis first
// 2. On miss, fetch from DB, store in cache, return
export async function getProduct(productId: string): Promise<Product> {
    const cacheKey = `product:${productId}`;

    // Step 1: Try cache first (Cache Hit path)
    const cached = await redis.get(cacheKey);
    if (cached) {
        const product: Product = JSON.parse(cached);
        console.log("[CACHE HIT]", productId);
        return product;
    }

    // Step 2: Cache Miss ,  fetch from database (expensive operation)
    console.log("[CACHE MISS] fetching from DB...", productId);
    const product = await fetchFromDatabase(productId); // simulate DB call

    // Step 3: Store in cache with a 1-hour TTL
    await redis.set(cacheKey, JSON.stringify(product), 'EX', 3600);

    return product;
}

// Write-Through: update DB and cache simultaneously
export async function updateProduct(product: Product): Promise<void> {
    // Step 1: Update in database
    await updateInDatabase(product);

    // Step 2: Write-through ,  update cache immediately
    const cacheKey = `product:${product.id}`;
    await redis.set(cacheKey, JSON.stringify(product), 'EX', 3600);

    console.log("[WRITE-THROUGH] DB + cache updated for", product.id);
}

// Mocks
async function fetchFromDatabase(id: string): Promise<Product> { return { id, name: "Product", price: 10 }; }
async function updateInDatabase(product: Product): Promise<void> {}

File: 9.Caching, the secret behind it all/code/rust/104_session_management_with_redis_python_1.rs
use redis::AsyncCommands;
use serde::{Deserialize, Serialize};

#[derive(Serialize, Deserialize, Debug)]
pub struct Product {
    pub id: String,
    pub name: String,
    pub price: f64,
}

// GetProduct implements Cache-Aside (Lazy) caching.
// 1. Check Redis first
// 2. On miss, fetch from DB, store in cache, return
pub async fn get_product(
    redis_conn: &mut redis::aio::Connection,
    product_id: &str,
) -> Result<Product, Box<dyn std::error::Error>> {
    let cache_key = format!("product:{}", product_id);

    // Step 1: Try cache first (Cache Hit path)
    let cached: Option<String> = redis_conn.get(&cache_key).await?;
    if let Some(cached_data) = cached {
        let product: Product = serde_json::from_str(&cached_data)?;
        println!("[CACHE HIT] {}", product_id);
        return Ok(product);
    }

    // Step 2: Cache Miss ,  fetch from database (expensive operation)
    println!("[CACHE MISS] fetching from DB... {}", product_id);
    let product = fetch_from_database(product_id).await?; // simulate DB call

    // Step 3: Store in cache with a 1-hour TTL
    let data = serde_json::to_string(&product)?;
    let _: () = redis_conn.set_ex(&cache_key, data, 3600).await?;

    Ok(product)
}

// Write-Through: update DB and cache simultaneously
pub async fn update_product(
    redis_conn: &mut redis::aio::Connection,
    product: &Product,
) -> Result<(), Box<dyn std::error::Error>> {
    // Step 1: Update in database
    update_in_database(product).await?;

    // Step 2: Write-through ,  update cache immediately
    let cache_key = format!("product:{}", product.id);
    let data = serde_json::to_string(product)?;
    let _: () = redis_conn.set_ex(&cache_key, data, 3600).await?;

    println!("[WRITE-THROUGH] DB + cache updated for {}", product.id);
    Ok(())
}

// Mocks
async fn fetch_from_database(id: &str) -> Result<Product, Box<dyn std::error::Error>> {
    Ok(Product { id: id.to_string(), name: "Product".to_string(), price: 10.0 })
}
async fn update_in_database(_product: &Product) -> Result<(), Box<dyn std::error::Error>> {
    Ok(())
}
